src/include/platform/DeviceInstanceInfoProvider.h

```diff
@@ -194,6 +194,19 @@ class DeviceInstanceInfoProvider
     {
         return CHIP_ERROR_NOT_IMPLEMENTED;
     }
+
+    /**
+     * @brief Obtain a Software Version String from the device's factory data.
+     *
+     * @param[in, out] buf Buffer to copy string.
+     *                     On CHIP_NO_ERROR return from this function this buffer will be null-terminated.
+     *                     On error CHIP_ERROR_BUFFER_TOO_SMALL there is no guarantee that buffer will be null-terminated.
+     * @param[in] bufSize Size of data, including the null terminator, that can be written to buf.
+     *                    This size should be +1 higher than maximum possible string.
+     * @returns CHIP_NO_ERROR on success, CHIP_ERROR_BUFFER_TOO_SMALL if the buffer was too small to fit string and null
+     * terminating. or another CHIP_ERROR from the underlying implementation if access fails.
+     */
+    virtual CHIP_ERROR GetSoftwareVersionString(char * buf, size_t bufSize) = 0;
```

Comment:
I know other APIs on here work like this, but please: use a MutableCharSpan outparam instead of the two separate things.

Reply:
Why is this moved here? The purpose of this class was to expose information that can vary across devices running the same software (hence, it typically needs to be read from a separate memory region that is not updated with OTA). Software version, on the other hand, is tied to the software (well except for testing scenarios in which you may want to override it at runtime).